02_cancer_type_classification/lasso_range_analysis/nbconverted/compare_optimizers_all.py

```diff
@@ -0,0 +1,318 @@
+# 
+# We'll do this for both CV (validation) datasets and test (holdout cancer type) datasets.
+
+# In[10]:
+
+
+# get mean AUPR values across folds/seeds
+ll_mean_aupr_df = (
+    all_perf_df[(all_perf_df.data_type == data_type) &
+                (all_perf_df.optimizer == 'liblinear')]
+      .drop(columns=['data_type', 'optimizer'])
+      .groupby(['gene', 'holdout_cancer_type', 'lasso_param'])
+      .agg(np.mean)
+      .reset_index()
+      .drop(columns=['seed', 'fold', 'auroc', 'nz_coefs'])
+)
+
+# get best LASSO parameter by mean AUPR, across all the ones we tried for this optimizer
+ll_max_lasso_ix = (ll_mean_aupr_df
+      .groupby(['gene', 'holdout_cancer_type'])
+      .aupr.idxmax()
+)
+ll_max_lasso_param_df = ll_mean_aupr_df.loc[ll_max_lasso_ix, :]
+
+print(ll_max_lasso_param_df.shape)
+ll_max_lasso_param_df.head(8)
+
+
+# In[11]:
+
+
+# get mean AUPR values across folds/seeds
+sgd_mean_aupr_df = (
+    all_perf_df[(all_perf_df.data_type == data_type) &
+                (all_perf_df.optimizer == 'sgd')]
+      .drop(columns=['data_type', 'optimizer'])
+      .groupby(['gene', 'holdout_cancer_type', 'lasso_param'])
+      .agg(np.mean)
+      .reset_index()
+      .drop(columns=['seed', 'fold', 'auroc', 'nz_coefs'])
+)
+
+# get best LASSO parameter by mean AUPR, across all the ones we tried for this optimizer
+sgd_max_lasso_ix = (sgd_mean_aupr_df
+      .groupby(['gene', 'holdout_cancer_type'])
+      .aupr.idxmax()
+)
+sgd_max_lasso_param_df = sgd_mean_aupr_df.loc[sgd_max_lasso_ix, :]
+
+print(sgd_max_lasso_param_df.shape)
+sgd_max_lasso_param_df.head(8)
+
+
+# In[12]:
+
+
+optimizer_diff_df = (ll_max_lasso_param_df
+    .merge(sgd_max_lasso_param_df,
+           left_on=['gene', 'holdout_cancer_type'],
+           right_on=['gene', 'holdout_cancer_type'])
+    .rename({'lasso_param_x': 'lasso_param_ll',
+             'lasso_param_y': 'lasso_param_sgd'})
+)
+optimizer_diff_df['ll_sgd_diff'] = (
+    optimizer_diff_df['aupr_x'] - optimizer_diff_df['aupr_y']
+)
+
+optimizer_diff_df.head()
+
+
+# In[13]:
+
+
+sns.set({'figure.figsize': (11, 6)})
+
+sns.histplot(optimizer_diff_df.ll_sgd_diff)
+plt.title(f'Distribution of (liblinear - SGD) performance differences, {data_type} data')
+plt.xlabel('AUPR(liblinear) - AUPR(SGD)')
+plt.gca().axvline(x=0, color='black', linestyle='--')
+
+
+# In[14]:
+
+
+# plot test performance vs. number of nonzero features
+sns.set({'figure.figsize': (28, 6)})
+sns.set_style('ticks')
+
+# order boxes by mean diff per gene
+gene_order = (optimizer_diff_df
+    .groupby('gene')
+    .agg(np.mean)
+    .sort_values(by='ll_sgd_diff', ascending=False)
+).index.values
+
+with sns.plotting_context('notebook', font_scale=1.5):
+    ax = sns.boxplot(data=optimizer_diff_df, order=gene_order, x='gene', y='ll_sgd_diff')
+    ax.axhline(0.0, linestyle='--', color='grey')
+    plt.xticks(rotation=90)
+    plt.title(f'Top {data_type} set performance difference between liblinear and SGD optimizers, per gene', y=1.02)
```

Comment:
Looks like you're looking at the top performing set (among all CV folds I think?) would you have different results if you compared all of them? Is it worth doing?

Reply:
Sorry, this isn't super clear - by "set" I mean holdout dataset, since I'm doing this both for "cv" which is the validation set with the same cancer types as the training data, and "test" which is the test/held out cancer type. Then the boxes are the median/IQR performance over all held-out cancer types and all CV folds for that gene (so I think I'm already doing what you're suggesting if I understand correctly).

The "top performing" part refers to the LASSO parameters - I'm not sure it would make a ton of sense to do a comparison over all LASSO parameters, since some of them perform really badly which would drag everything toward 0.

Maybe calling it a "dataset" instead of "set" would make this title a bit clearer. I'll also put the explanation above in the notebook for future reference.